Source/WebCore/page/text-extraction/TextExtraction.cpp

```diff
@@ -0,0 +1,189 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "TextExtraction.h"
+
+#include "ComposedTreeIterator.h"
+#include "HTMLBodyElement.h"
+#include "HTMLButtonElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLInputElement.h"
+#include "Page.h"
+#include "RenderBox.h"
+#include "RenderLayer.h"
+#include "RenderLayerModelObject.h"
+#include "RenderLayerScrollableArea.h"
+#include "SimpleRange.h"
+#include "Text.h"
+#include "TextIterator.h"
+
+namespace WebCore {
+namespace TextExtraction {
+
+static inline HashMap<Ref<Text>, String> collectText(ContainerNode& container)
+{
+    HashMap<Ref<Text>, String> nodeToTextMap;
+    RefPtr<Text> lastTextNode;
+    StringBuilder textForLastTextNode;
+    for (TextIterator iterator { makeRangeSelectingNodeContents(container) }; !iterator.atEnd(); iterator.advance()) {
+        if (iterator.text().isEmpty())
+            continue;
+
+        RefPtr textNode = dynamicDowncast<Text>(iterator.node());
+        if (!textNode)
+            continue;
+
+        if (!lastTextNode)
+            lastTextNode = textNode;
+
+        if (lastTextNode == textNode) {
+            textForLastTextNode.append(iterator.text());
+            continue;
+        }
+
+        if (auto text = textForLastTextNode.toString().trim(isASCIIWhitespace<UChar>); !text.isEmpty())
+            nodeToTextMap.add(*lastTextNode, textForLastTextNode.toString());
+
+        textForLastTextNode.clear();
+        textForLastTextNode.append(iterator.text());
+        lastTextNode = textNode;
+    }
+
+    if (auto text = textForLastTextNode.toString().trim(isASCIIWhitespace<UChar>); lastTextNode && !text.isEmpty())
+        nodeToTextMap.add(*lastTextNode, textForLastTextNode.toString());
+
+    return nodeToTextMap;
+}
+
+static bool shouldIncludeChildren(const ItemData& data)
+{
+    return WTF::switchOn(data,
+        [&](const TextItemData&) { return false; },
+        [&](const ScrollableItemData&) { return true; },
+        [&](const EditableItemData&) { return true; },
+        [&](const ImageItemData&) { return false; },
+        [&](const InteractiveItemData&) { return true; },
+        [&](ContainerType) { return true; }
+    );
+}
+
+static inline FloatRect rootViewBounds(Node& node)
+{
+    auto view = node.document().protectedView();
+    if (UNLIKELY(!view))
+        return { };
+
+    if (!node.renderer())
+        return { };
+
+    return view->contentsToRootView(node.renderer()->absoluteBoundingBoxRect());
+}
+
+static inline std::optional<ItemData> extractItemData(Node& node, const HashMap<Ref<Text>, String>& extractedText)
+{
+    if (!node.renderer())
+        return std::nullopt;
```

Comment:
Should we exclude items with `visibility: hidden` or `opacity: 0`?

Reply:
Hm... I think it makes sense to skip `visibility: hidden` elements, but for `opacity: 0` I think it's a bit difficult to say (since the text can still be selected and copied). I'll just bail on hidden for the time being.

Thanks for the review!